### DISCUSSION TITLE
Principle: Governance through git

### DISCUSSION THREAD

**Discussion by moshloop** [open-gitops/project#82] [General]
### 5. Governance through git

<div style="background: #E3F2FD; border-left: 3px solid #0D47A1; padding:10px; margin-bottom:5px; font-style:italic">
In the complex and adapative socio-technical system in which GitOps is most often applied,  the process and decision making criteria that lead to a desired state, is often as or more important than the state itself.
</div>

**Comment by lloydchang** [comment]
The objective of GitOps is convergence to a declarative state.

While Git facilitates GitOps, Git is optional.

This relates to https://github.com/gitops-working-group/gitops-working-group/discussions/76#discussioncomment-633924

Thank you @moshloop @o6uoq

**Reply by moshloop** [reply]
This might be better phrased as *Governenace through pull/merge request" - Just applying changes into a state store isn't enough, the rationale, policy and signals that led to a state change being committed are as important (or even more) than the current desired state

**Comment by scottrigby** [comment]
I personally agree with your statement above. However I believe this is an extension of principle 2 in the revised principles draft – of which auditability is one benefit, and audit requirements are an implementation detail: 

>2. [**The principle of immutable desired state versions**]()
>
>    _Desired State_ is stored in a way that supports versioning, immutability of versions, and retains a complete version history.

I would suggest that statements like this be proposed as part of the extended notes for principle 2.

**Reply by lloydchang** [reply]
This relates to https://github.com/gitops-working-group/gitops-working-group/pull/48#discussion_r617923248

**Reply by moshloop** [reply]
I don't agree - If I push straight to master, and never rewrite history I still have immutable desired state versions. 

It is not only the desired state that needs to be immutable but also the decision-making process that should a) be recorded and b) immutable

**Reply by lloydchang** [reply]
@moshloop That seems onerous. [Heroku](https://blog.heroku.com/api_and_external_git_access) and [Flagger](https://github.com/stefanprodan/gitops-appmesh/commit/73b944b5af32bfbba5eaf531dfaf12f2486f0915#diff-b335630551682c19a781afebcf4d07bf978fb1f8ac04c6bf87428ed5106870f5R160) allow `git push` without any merge requests nor pull requests at all; [I consider them as GitOps](https://github.com/gitops-working-group/gitops-working-group/discussions/53#discussioncomment-638271).
